File: backend/app/services/haversine.py
import math
from typing import List, Dict, Tuple

EARTH_RADIUS_KM = 6371  # Earth's radius in kilometers

def haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate the Haversine distance between two geographical points
    Returns distance in kilometers
    """
    # Convert to radians
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    delta_lat = math.radians(lat2 - lat1)
    delta_lon = math.radians(lon2 - lon1)
    
    # Haversine formula
    a = (math.sin(delta_lat / 2) ** 2 +
         math.cos(lat1_rad) * math.cos(lat2_rad) *
         math.sin(delta_lon / 2) ** 2)
    
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    distance = EARTH_RADIUS_KM * c
    
    return distance

def calculate_geographic_centroid(points: List[Dict[str, float]]) -> Dict[str, float]:
    """
    Calculate the geographic centroid of multiple points
    Uses spherical geometry for accuracy
    """
    if not points:
        raise ValueError("Cannot calculate centroid of empty array")
    
    if len(points) == 1:
        return {"latitude": points[0]["latitude"], "longitude": points[0]["longitude"]}
    
    # Convert to Cartesian coordinates
    x, y, z = 0, 0, 0
    
    for point in points:
        lat_rad = math.radians(point["latitude"])
        lon_rad = math.radians(point["longitude"])
        
        x += math.cos(lat_rad) * math.cos(lon_rad)
        y += math.cos(lat_rad) * math.sin(lon_rad)
        z += math.sin(lat_rad)
    
    total = len(points)
    x /= total
    y /= total
    z /= total
    
    # Convert back to latitude/longitude
    central_longitude = math.atan2(y, x)
    central_square_root = math.sqrt(x * x + y * y)
    central_latitude = math.atan2(z, central_square_root)
    
    return {
        "latitude": math.degrees(central_latitude),
        "longitude": math.degrees(central_longitude)
    }

def calculate_density_within_radius(
    center: Dict[str, float],
    points: List[Dict[str, float]],
    radius_km: float
) -> int:
    """
    Calculate the number of points within a given radius
    """
    count = 0
    for point in points:
        distance = haversine_distance(
            center["latitude"],
            center["longitude"],
            point["latitude"],
            point["longitude"]
        )
        if distance <= radius_km:
            count += 1
    return count

def find_nearest_point(
    location: Dict[str, float],
    points: List[Dict[str, float]]
) -> Tuple[Dict[str, float], float]:
    """
    Find the nearest point to a given location
    Returns (nearest_point, distance_km)
    """
    if not points:
        return None, float('inf')
    
    nearest_point = points[0]
    min_distance = haversine_distance(
        location["latitude"],
        location["longitude"],
        points[0]["latitude"],
        points[0]["longitude"]
    )
    
    for point in points[1:]:
        distance = haversine_distance(
            location["latitude"],
            location["longitude"],
            point["latitude"],
            point["longitude"]
        )
        if distance < min_distance:
            min_distance = distance
            nearest_point = point
    
    return nearest_point, min_distance

def get_points_within_radius(
    center: Dict[str, float],
    points: List[Dict[str, float]],
    radius_km: float
) -> List[Tuple[Dict[str, float], float]]:
    """
    Get all points within a radius sorted by distance
    Returns list of (point, distance) tuples
    """
    result = []
    for point in points:
        distance = haversine_distance(
            center["latitude"],
            center["longitude"],
            point["latitude"],
            point["longitude"]
        )
        if distance <= radius_km:
            result.append((point, distance))
    
    # Sort by distance
    result.sort(key=lambda x: x[1])
    return result

File: frontend/utils/haversine.ts
// Haversine Distance Formula for accurate geographical distance calculation
// Returns distance in kilometers

export interface GeoPoint {
    latitude: number;
    longitude: number;
}

const EARTH_RADIUS_KM = 6371; // Earth's radius in kilometers

/**
 * Calculate the Haversine distance between two geographical points
 * @param point1 First geographical point
 * @param point2 Second geographical point
 * @returns Distance in kilometers
 */
export function haversineDistance(point1: GeoPoint, point2: GeoPoint): number {
    const lat1Rad = toRadians(point1.latitude);
    const lat2Rad = toRadians(point2.latitude);
    const deltaLatRad = toRadians(point2.latitude - point1.latitude);
    const deltaLonRad = toRadians(point2.longitude - point1.longitude);

    const a =
        Math.sin(deltaLatRad / 2) * Math.sin(deltaLatRad / 2) +
        Math.cos(lat1Rad) *
        Math.cos(lat2Rad) *
        Math.sin(deltaLonRad / 2) *
        Math.sin(deltaLonRad / 2);

    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    const distance = EARTH_RADIUS_KM * c;

    return distance;
}

/**
 * Convert degrees to radians
 */
function toRadians(degrees: number): number {
    return degrees * (Math.PI / 180);
}

/**
 * Convert radians to degrees
 */
function toDegrees(radians: number): number {
    return radians * (180 / Math.PI);
}

/**
 * Calculate density within a radius using Haversine distance
 * @param center Center point
 * @param points Array of points to check
 * @param radiusKm Radius in kilometers
 * @returns Number of points within the radius
 */
export function calculateDensityWithinRadius(
    center: GeoPoint,
    points: GeoPoint[],
    radiusKm: number
): number {
    return points.filter(point => {
        const distance = haversineDistance(center, point);
        return distance <= radiusKm;
    }).length;
}

/**
 * Find the nearest point to a given location
 * @param location Target location
 * @param points Array of points to search
 * @returns Object with nearest point and distance in km
 */
export function findNearestPoint(
    location: GeoPoint,
    points: GeoPoint[]
): { point: GeoPoint; distance: number } | null {
    if (points.length === 0) return null;

    let nearestPoint = points[0];
    let minDistance = haversineDistance(location, points[0]);

    for (let i = 1; i < points.length; i++) {
        const distance = haversineDistance(location, points[i]);
        if (distance < minDistance) {
            minDistance = distance;
            nearestPoint = points[i];
        }
    }

    return {
        point: nearestPoint,
        distance: minDistance,
    };
}

/**
 * Calculate the centroid of multiple geographical points
 * Using spherical geometry for accuracy
 */
export function calculateGeographicCentroid(points: GeoPoint[]): GeoPoint {
    if (points.length === 0) {
        throw new Error("Cannot calculate centroid of empty array");
    }

    if (points.length === 1) {
        return points[0];
    }

    // Convert to Cartesian coordinates
    let x = 0, y = 0, z = 0;

    for (const point of points) {
        const latRad = toRadians(point.latitude);
        const lonRad = toRadians(point.longitude);

        x += Math.cos(latRad) * Math.cos(lonRad);
        y += Math.cos(latRad) * Math.sin(lonRad);
        z += Math.sin(latRad);
    }

    const total = points.length;
    x /= total;
    y /= total;
    z /= total;

    // Convert back to latitude/longitude
    const centralLongitude = Math.atan2(y, x);
    const centralSquareRoot = Math.sqrt(x * x + y * y);
    const centralLatitude = Math.atan2(z, centralSquareRoot);

    return {
        latitude: toDegrees(centralLatitude),
        longitude: toDegrees(centralLongitude),
    };
}

/**
 * Get all points within a radius sorted by distance
 */
export function getPointsWithinRadius(
    center: GeoPoint,
    points: GeoPoint[],
    radiusKm: number
): Array<{ point: GeoPoint; distance: number }> {
    return points
        .map(point => ({
            point,
            distance: haversineDistance(center, point),
        }))
        .filter(item => item.distance <= radiusKm)
        .sort((a, b) => a.distance - b.distance);
}
